vowpalwabbit/offset_tree.cc

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "offset_tree.h"
+#include "parse_args.h"  // setup_base()
+#include "learner.h"     // init_learner()
+#include "action_score.h"
+
+using namespace VW::config;
+using namespace VW::LEARNER;
+
+namespace VW { namespace offset_tree {
+
+  tree_node::tree_node(uint32_t node_id, uint32_t left_node_id, uint32_t right_node_id, uint32_t p_id, bool is_leaf)
+    : id(node_id), left_id(left_node_id), right_id(right_node_id), parent_id(p_id), is_leaf(is_leaf)
+  {
+  }
+
+  bool tree_node::operator==(const tree_node& rhs) const
+  {
+    if (this == &rhs)
+      return true;
+    return (id == rhs.id && left_id == rhs.left_id && right_id == rhs.right_id && is_leaf == rhs.is_leaf && parent_id == rhs.parent_id);
+  }
+
+  bool tree_node::operator!=(const tree_node& rhs) const { return !(*this == rhs); }
+
+  void min_depth_binary_tree::build_tree(uint32_t num_nodes)
+  {
+    // Sanity checks
+    if (_initialized)
+    {
```

Comment:
When would it make sense to call `build_tree` again even if the number of nodes are the same?

Reply:
I don't see a use case for that.  Instead of the additional complexity/cost of tracking and reversing changes, it seems better to create a new tree.